File: app/src/main/java/cn/alvkeke/dropto/ui/activity/ShareRecvActivity.java
package cn.alvkeke.dropto.ui.activity;

import android.content.ComponentName;
import android.content.Intent;
import android.net.Uri;
import android.os.Build;
import android.os.Bundle;
import android.util.Log;
import android.widget.Toast;

import androidx.activity.EdgeToEdge;
import androidx.annotation.NonNull;
import androidx.appcompat.app.AppCompatActivity;

import java.io.File;
import java.util.ArrayList;

import cn.alvkeke.dropto.data.Category;
import cn.alvkeke.dropto.mgmt.Global;
import cn.alvkeke.dropto.data.ImageFile;
import cn.alvkeke.dropto.data.NoteItem;
import cn.alvkeke.dropto.service.CoreService;
import cn.alvkeke.dropto.service.CoreServiceConnection;
import cn.alvkeke.dropto.service.Task;
import cn.alvkeke.dropto.storage.DataLoader;
import cn.alvkeke.dropto.storage.FileHelper;
import cn.alvkeke.dropto.ui.fragment.CategorySelectorFragment;

public class ShareRecvActivity extends AppCompatActivity
        implements CategorySelectorFragment.CategorySelectListener,
        Task.ResultListener {

    private CoreService service = null;
    private final CoreServiceConnection serviceConn = new CoreServiceConnection(this) {
        @Override
        public void execOnServiceConnected(ComponentName componentName, Bundle bundleAfterConnected) {
            service = getService();
        }

        @Override
        public void execOnServiceDisconnected() {
            super.execOnServiceDisconnected();
            service = null;
        }
    };

    private void setupCoreService() {
        Intent serviceIntent = new Intent(this, CoreService.class);
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
            startForegroundService(serviceIntent);
        } else {
            startService(serviceIntent);
        }
        bindService(serviceIntent, serviceConn, BIND_AUTO_CREATE);
    }

    private void clearCoreService() {
        if (service == null) return;
        unbindService(serviceConn);
    }

    @Override
    protected void onDestroy() {
        super.onDestroy();
        clearCoreService();
    }

    private CategorySelectorFragment categorySelectorFragment;
    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        EdgeToEdge.enable(this);

        if (savedInstanceState != null) {
            this.finish();
            // just end after change to dark mode
            return;
        }

        setupCoreService();
        Intent intent = getIntent();
        String action = intent.getAction();
        if (action == null) {
            Toast.makeText(this, "action is null", Toast.LENGTH_SHORT).show();
            finish();
            return;
        }

        if (!isActionAllow(action)) {
            Toast.makeText(this, "unknown action: " + action, Toast.LENGTH_SHORT).show();
            finish();
        }

        pendingText = handleTextInfo(intent);
        String type = intent.getType();
        if (type != null) {
            if (type.startsWith("image/")) {
                pendingUris = handleImageUris(intent);
            }
        }

        categorySelectorFragment = new CategorySelectorFragment();
        categorySelectorFragment.setCategories(DataLoader.getCategories());
        getSupportFragmentManager().beginTransaction()
                .add(categorySelectorFragment, null)
                .commit();
    }

    private boolean isActionAllow(String action) {
        switch (action) {
            case Intent.ACTION_SEND:
            case Intent.ACTION_SEND_MULTIPLE:
                return true;
        }
        return false;
    }

    private String pendingText = null;
    private ArrayList<Uri> pendingUris = null;
    private boolean noPendingUris() {
        return pendingUris == null || pendingUris.isEmpty();
    }

    @Override
    public void onSelected(int index, Category category) {
        if (pendingText.isEmpty() && noPendingUris()) {
            Toast.makeText(this, "Empty item, abort", Toast.LENGTH_SHORT).show();
            finish();
            return;
        }

        NoteItem recvNote = new NoteItem(pendingText);
        recvNote.categoryId = category.id;
        if (pendingUris != null) {
            for (Uri uri : pendingUris) {
                ImageFile imageFile = extraImageFileFromUri(uri);
                if (imageFile != null)
                    recvNote.addImageFile(imageFile);
            }
        }
        service.queueTask(Task.createNote(recvNote, null));
        finish();
    }

    @Override
    public void onError(String error) {
        Toast.makeText(ShareRecvActivity.this, error, Toast.LENGTH_SHORT).show();
        finish();
    }

    @NonNull
    private String handleTextInfo(@NonNull Intent intent) {
        String text = intent.getStringExtra(Intent.EXTRA_TEXT);
        if (text == null) {
            Log.e(this.toString(), "Failed to get shared text");
            return "";
        }
        return text.trim();
    }

    private ArrayList<Uri> handleImageUris(@NonNull Intent intent) {
        ArrayList<Uri> uris = new ArrayList<>();
        String action = intent.getAction();
        if (action == null)
            return null;

        switch (action) {
            case Intent.ACTION_SEND:
                uris.add(handleSingleImageUri(intent));
                break;
            case Intent.ACTION_SEND_MULTIPLE:
                handleMultipleImageUris(intent, uris);
                break;
        }

        return uris;
    }

    private Uri handleSingleImageUri(@NonNull Intent intent) {
        return intent.getParcelableExtra(Intent.EXTRA_STREAM);
    }

    private void handleMultipleImageUris(@NonNull Intent intent, ArrayList<Uri> uris) {
        ArrayList<Uri> imageUris = intent.getParcelableArrayListExtra(Intent.EXTRA_STREAM);
        if (imageUris == null || imageUris.isEmpty()) return;
        uris.addAll(imageUris);
    }

    private ImageFile extraImageFileFromUri(Uri uri) {
        File folder = Global.getFolderImage(this);
        File md5file = FileHelper.saveUriToFile(this, uri, folder);
        if (md5file == null) return null;
        String imgName = FileHelper.getFileNameFromUri(this, uri);
        return new ImageFile(md5file, imgName);
    }

    private void onCategoryTaskFinish(Task task) {
        if (task.result < 0) return;
        if (categorySelectorFragment == null) return;
        switch (task.job) {
            case CREATE:
            case REMOVE:
            case UPDATE:
                categorySelectorFragment.notifyItemListChanged(
                        Task.jobToNotify(task.job), task.result, (Category) task.param);
                break;
            default:
                break;
        }
    }

    @Override
    public void onTaskFinish(Task task, Object param) {
        if (task.type == Task.Type.Category)
            onCategoryTaskFinish(task);
    }
}

File: app/src/main/java/cn/alvkeke/dropto/ui/activity/ShareRecvActivity.kt
package cn.alvkeke.dropto.ui.activity

import android.content.Intent
import android.net.Uri
import android.os.Bundle
import android.util.Log
import android.widget.Toast
import androidx.activity.enableEdgeToEdge
import androidx.appcompat.app.AppCompatActivity
import cn.alvkeke.dropto.DroptoApplication
import cn.alvkeke.dropto.data.AttachmentFile
import cn.alvkeke.dropto.data.AttachmentFile.Type
import cn.alvkeke.dropto.data.Category
import cn.alvkeke.dropto.data.NoteItem
import cn.alvkeke.dropto.storage.DataLoader.categories
import cn.alvkeke.dropto.storage.FileHelper
import cn.alvkeke.dropto.storage.FileHelper.getFileNameFromUri
import cn.alvkeke.dropto.storage.FileHelper.saveUriToFile
import cn.alvkeke.dropto.ui.fragment.CategorySelectorFragment

class ShareRecvActivity : AppCompatActivity(), CategorySelectorFragment.CategorySelectListener {

    private val app: DroptoApplication
        get() = application as DroptoApplication

    private val categorySelectorFragment: CategorySelectorFragment by lazy {
        CategorySelectorFragment()
    }

    private lateinit var senderPackageName: String
    private fun Intent.tryGetSenderPackageName(): String {
        callingPackage?.let {
            Log.d(TAG, "get sender package name from callingPackage: $it")
            return it
        }

        callingActivity?.packageName?.let {
            Log.d(TAG, "get sender package name from callingActivity: $it")
            return it
        }

        getPackageFromReferrerUri(referrer)?.let {
            Log.d(TAG, "get sender package name from getReferrer: $it")
            return it
        }

        intent.getStringExtra(Intent.EXTRA_REFERRER_NAME)?.let {
            val fromExtra = getPackageFromReferrerName(it)
            if (fromExtra != null) {
                Log.d(TAG, "get sender package name from EXTRA_REFERRER_NAME: $fromExtra")
                return fromExtra
            }
        }

        intent.getStringExtra(Intent.EXTRA_PACKAGE_NAME)?.let {
            Log.d(TAG, "get sender package name from EXTRA_PACKAGE_NAME: $it")
            return it
        }

        intent.`package`?.let {
            Log.d(TAG, "get sender package name from intent.package: $it")
            return it
        }

        this.getContentAuthorityFromIntent()?.let {
            Log.d(TAG, "get sender content authority from intent: $it")
            return it
        }

        return UNKNOWN_SENDER_PACKAGE
    }

    private fun getPackageFromReferrerUri(referrer: Uri?): String? {
        if (referrer == null) return null
        if (referrer.scheme == "android-app") return referrer.host
        return referrer.authority ?: referrer.host
    }

    private fun getPackageFromReferrerName(referrerName: String): String? {
        if (referrerName.isBlank()) return null
        return try {
            val parsed = Uri.parse(referrerName)
            when (parsed.scheme) {
                null -> referrerName
                "android-app" -> parsed.host
                else -> parsed.authority ?: parsed.host
            }
        } catch (ex: Exception) {
            Log.d(TAG, "Failed to parse EXTRA_REFERRER_NAME: $referrerName", ex)
            referrerName
        }
    }

    private fun Intent.getContentAuthorityFromIntent(): String? {
        val clipData = clipData
        if (clipData != null && clipData.itemCount > 0) {
            val uri = clipData.getItemAt(0)?.uri
            if (uri != null && uri.scheme == "content") return uri.authority
        }

        val streamUri = getParcelableExtra(Intent.EXTRA_STREAM, Uri::class.java)
        if (streamUri != null && streamUri.scheme == "content") return streamUri.authority

        return null
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        this.enableEdgeToEdge()

        val intent = getIntent()
        senderPackageName = intent.tryGetSenderPackageName()
        val action = intent.action
        if (action == null) {
            Toast.makeText(this, "action is null", Toast.LENGTH_SHORT).show()
            finish()
            return
        }

        if (!isActionAllow(action)) {
            Toast.makeText(this, "unknown action: $action", Toast.LENGTH_SHORT).show()
            finish()
        }

        pendingText = handleTextInfo(intent)
        val type = intent.type
        pendingType = intent.type ?: "*/*"
        Log.e(TAG, "received type: $type")
        pendingUris = handleReceivedUris(intent)

        categorySelectorFragment.prepare(categories, this)
        supportFragmentManager.beginTransaction()
            .add(categorySelectorFragment, null)
            .commit()
    }

    private fun isActionAllow(action: String): Boolean {
        when (action) {
            Intent.ACTION_SEND, Intent.ACTION_SEND_MULTIPLE -> return true
        }
        return false
    }

    private lateinit var pendingText: String
    private lateinit var pendingUris: ArrayList<Uri>
    private lateinit var pendingType: String

    override fun onSelected(index: Int, category: Category) {
        if (pendingText.isEmpty() && pendingUris.isEmpty()) {
            Toast.makeText(this, "Empty item, abort", Toast.LENGTH_SHORT).show()
            finish()
            return
        }

        Log.v(TAG, "selected category: ${category.title}")

        val recvNote = NoteItem(pendingText, senderPackageName)
        recvNote.categoryId = category.id
        for (uri in pendingUris) {
            val imageFile = extractAttachmentFromUri(uri)
            if (imageFile == null) {
                Log.e(this.toString(), "Failed to extract attachment from uri: $uri")
                continue
            }
            recvNote.attachments.add(imageFile)
        }

        if (recvNote.attachments.size != pendingUris.size) {
            val message = "Failed to create note, Some or all shared files cannot be saved"
            Log.e(TAG, message + ", ${recvNote.attachments.size} / ${pendingUris.size}")
            Toast.makeText(this, message, Toast.LENGTH_SHORT).show()
        } else if (recvNote.attachments.isEmpty() && recvNote.text.isEmpty()) {
            val message = "Failed to create note, no available content"
            Log.e(TAG, message)
            Toast.makeText(this, message, Toast.LENGTH_SHORT).show()
        } else {
            app.service?.createNote(recvNote)
        }
        finish()
    }

    override fun onCancel() {
        finish()
    }

    private fun handleTextInfo(intent: Intent): String {
        val text = intent.getStringExtra(Intent.EXTRA_TEXT)
        if (text == null) {
            Log.e(this.toString(), "Failed to get shared text")
            return ""
        }
        return text.trim { it <= ' ' }
    }

    private fun handleReceivedUris(intent: Intent): ArrayList<Uri> {
        return when (intent.action) {
            Intent.ACTION_SEND -> {
                val uris = ArrayList<Uri>()
                val uri = intent.getParcelableExtra(
                    Intent.EXTRA_STREAM,
                    Uri::class.java)
                if (uri != null) uris.add(uri)
                uris
            }
            Intent.ACTION_SEND_MULTIPLE -> {
                intent.getParcelableArrayListExtra(
                    Intent.EXTRA_STREAM,
                    Uri::class.java)?: ArrayList()
            }
            else -> error("action not allowed: ${intent.action}, should not reach here")
        }
    }

    private fun extractAttachmentFromUri(uri: Uri): AttachmentFile? {
        val folder = FileHelper.attachmentStorage
        val md5file = saveUriToFile(this, uri, folder)
        if (md5file == null) {
            Log.e(this.toString(), "Failed to save uri to file: $uri")
            return null
        }
        val fileName = getFileNameFromUri(this, uri)
        val mimeType = FileHelper.mimeTypeFromFileName(fileName)
        val attachmentType = if (mimeType.startsWith("image/") ||
            mimeType.startsWith("video/")) {
            Type.MEDIA
        } else {
            Type.FILE
        }

        return AttachmentFile(md5file, fileName, attachmentType)
    }

    companion object {
        private const val TAG = "ShareRecvActivity"

        const val UNKNOWN_SENDER_PACKAGE = "UNKNOWN"
    }

}